staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go

```diff
@@ -17,68 +17,47 @@ limitations under the License.
 package configuration
 
 import (
-	"fmt"
-	"reflect"
-
-	"github.com/golang/glog"
+	"sort"
 
 	"k8s.io/api/admissionregistration/v1beta1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/labels"
+	admissionregistrationlisters "k8s.io/client-go/listers/admissionregistration/v1beta1"
 )
 
-type ValidatingWebhookConfigurationLister interface {
-	List(opts metav1.ListOptions) (*v1beta1.ValidatingWebhookConfigurationList, error)
-}
-
 // ValidatingWebhookConfigurationManager collects the validating webhook objects so that they can be called.
 type ValidatingWebhookConfigurationManager struct {
-	*poller
+	lister admissionregistrationlisters.ValidatingWebhookConfigurationLister
 }
 
-func NewValidatingWebhookConfigurationManager(c ValidatingWebhookConfigurationLister) *ValidatingWebhookConfigurationManager {
-	getFn := func() (runtime.Object, error) {
-		list, err := c.List(metav1.ListOptions{})
-		if err != nil {
-			if errors.IsNotFound(err) || errors.IsForbidden(err) {
-				glog.V(5).Infof("ValidatingWebhookConfiguration are disabled due to an error: %v", err)
-				return nil, ErrDisabled
-			}
-			return nil, err
-		}
-		return mergeValidatingWebhookConfigurations(list), nil
-	}
-
+func NewValidatingWebhookConfigurationManager(lister admissionregistrationlisters.ValidatingWebhookConfigurationLister) *ValidatingWebhookConfigurationManager {
 	return &ValidatingWebhookConfigurationManager{
-		newPoller(getFn),
+		lister: lister,
 	}
 }
 
 // Webhooks returns the merged ValidatingWebhookConfiguration.
 func (im *ValidatingWebhookConfigurationManager) Webhooks() (*v1beta1.ValidatingWebhookConfiguration, error) {
-	configuration, err := im.poller.configuration()
+	configurations, err := im.lister.List(labels.Everything())
```

Comment:
because this is called on every request to the API server, we shouldn't be doing transformation/allocation in this path

1. I'd expect to maintain the configuration, sorted and merged, and update that on every change notification from the informer
2. continue returning an error from Webhooks() until we've had a successful sync

See https://github.com/liggitt/kubernetes/commit/a0da6ada1bf9227719c4014c5d0690674801ef87 as an example for one of the managers

Reply:
Thanks for your example.  I also thought about this as cm, but referenced rbac I abandoned. And the disadvantage you say really exist. I will fix it now.